### ISSUE TITLE
Setup xfuncname for elisp and other filetypes

### ISSUE THREAD

**Issue by alphapapa** [magit/magit#3706]
By default, git's `funcname` regexp, which is responsible for determining the containing function for diff hunk context, does not work with elisp.  This can be fixed by:

1.  Add to `repo/.gitattributes`:
```
*.el diff=elisp
```
2.  Add to `repo/.git/config`:
```ini
[diff "elisp"]
      xfuncname = ^\\(def.*$
```

Emacs lisp is not the only language that may need this kind of change, but it's especially helpful, as it adds `(defun ...` headers to diff hunks.

It seems like it would be a useful feature for Magit to have a way to easily adjust these settings, perhaps with a few common filetypes built-in, so that with a few keypresses, a repo could be configured for elisp.

What do you think?  If this would be a useful addition, I might be able to work on a PR, although it would probably be good to discuss some basic design first, so that whatever I come up with would be "idiomatically magit".  :)

Thanks for your work on Magit!

**Comment by basil-conto** [comment]
See also the corresponding configuration in the Emacs source tree itself: search for "xfuncname" in [`autogen.sh`](http://git.savannah.gnu.org/cgit/emacs.git/tree/autogen.sh).

**Comment by tarsius** [comment]
I think adding some sort of setup wizard is not best way.  Users who don't know about this feature won't go looking for it and users who know about it but need help setting it up will instead go looking for the documentation.

So I propose we improve the documentation instead.  Personally I would go straight for the Git documentation, but improving the Magit documentation might help some users.

This is what I think we should do:

1. Write a text that explains how the feature works in Git and how Magit uses that. Additionally it should provide some examples and link to the relevant Git documentation.
2. Add the text to the manual and/or wiki.
3. Publish a blog post, which could be identical or just similar to that text.
4. Locate the relevant doc-strings and parts of the manual and link to that text.

Another reason why I don't think any *code* should be added to Magit is that I think each user should use it at most one time.  This should be configured once per language in a global configuration file.

So in summary, I propose that we improve the documentation and make one or more public service announcement, instead of adding code.

**Comment by phil-s** [comment]
Is there an upstream git issue for this?  It would be good if future git versions supported elisp OOTB.

**Comment by kyleam** [comment]
> Is there an upstream git issue for this?

Searching "lisp xfuncname" at https://public-inbox.org/git/ doesn't suggest that any requests have been made or patches submitted.

>  It would be good if future git versions supported elisp OOTB.

I think it'd be good to add a lisp pattern to userdiff.c, though I believe the user would still need to add the gitattributes line.

For anyone is interested in submitting a patch, here's an example of a patch that added support for golang: https://public-inbox.org/git/20180301111907.17607-1-alban.gruin@gmail.com/

**Comment by tarsius** [comment]
Closing since I am not gonna add any code.

**Comment by tarsius** [comment]
I've added this to https://github.com/magit/magit/wiki/Important-Git-settings.
